contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1024 @@
+    address tokenHolder;
+    uint256 scheduleTokenId;
+    uint256 startTime;
+    uint256 endTime;
+    uint256 balance;
+    uint256 claimableAmount;
+    uint256 claimedAmount;
+    uint256 quantityRevoked;
+    bool exists;
+  }
+
+  struct RestrictionDuration {
+    uint256 duration;
+    bool wasSet;
+  }
+
+  bytes32 public constant PAUSER_ROLE = keccak256("PAUSER_ROLE");
+
+  /**
+   * @notice Role conferring creation of schedules.
+   *
+   * @dev the Removal contract is granted this role during this contract's deployment.
+   */
+  bytes32 public constant SCHEDULE_CREATOR_ROLE =
+    keccak256("SCHEDULE_CREATOR_ROLE");
+
+  /**
+   * @notice Role conferring revocation of restricted tokens.
+   *
+   * @dev only Nori admin address should have this role.
+   */
+  bytes32 public constant TOKEN_REVOKER_ROLE = keccak256("TOKEN_REVOKER_ROLE");
+
+  /**
+   * @notice Used to register the ERC777TokensRecipient recipient interface in the
+   * ERC-1820 registry
+   *
+   * @dev Registering that RestrictedNORI implements the ERC777TokensRecipient interface with the registry is a
+   * requirement to be able to receive ERC-777 BridgedPolygonNORI tokens. Once registered, sending BridgedPolygonNORI
+   * tokens to this contract will trigger tokensReceived as part of the lifecycle of the BridgedPolygonNORI transaction
+   */
+  bytes32 public constant ERC777_TOKENS_RECIPIENT_HASH =
+    keccak256("ERC777TokensRecipient");
+
+  mapping(uint256 => mapping(uint256 => RestrictionDuration))
+    private _methodologyAndVersionToScheduleDuration;
+
+  mapping(address => EnumerableSetUpgradeable.UintSet)
+    private _addressToScheduleIdSet;
+
+  mapping(uint256 => Schedule) private _scheduleIdToScheduleStruct;
+
+  EnumerableSetUpgradeable.UintSet private _allScheduleIds;
+
+  /**
+   * @notice The BridgedPolygonNORI contract for which this contract wraps tokens.
+   */
+  FIFOMarket private _market;
+
+  /**
+   * @notice The BridgedPolygonNORI contract for which this contract wraps tokens.
+   */
+  BridgedPolygonNORI private _bridgedPolygonNori;
+
+  /**
+   * @notice The Removal contract that accounts for carbon removal supply.
+   */
+  Removal private _removal;
+
+  /**
+   * @notice The [ERC-1820](https://eips.ethereum.org/EIPS/eip-1820) pseudo-introspection registry
+   * contract
+   *
+   * @dev Registering that RestrictedNORI implements the ERC777TokensRecipient interface with the registry is a
+   * requirement to be able to receive ERC-777 BridgedPolygonNORI tokens. Once registered, sending BridgedPolygonNORI
+   * tokens to this contract will trigger tokensReceived as part of the lifecycle of the BridgedPolygonNORI transaction
+   */
+  IERC1820RegistryUpgradeable internal constant _ERC1820_REGISTRY =
+    IERC1820RegistryUpgradeable(0x1820a4B7618BdE71Dce8cdc73aAB6C95905faD24);
+
+  /**
+   * @notice Emitted on successful creation of a new schedule.
+   */
+  event ScheduleCreated(uint256 indexed projectId);
```

Comment:
i'm going to leave `totalSupply` out because it should always be 0 when a schedule is first created (because it happens when removals are listed and therefore have not been sold yet), but I will include the other two pieces of info